lib/Target/DebugInfo/EmitHGLDD.cpp

```diff
@@ -79,26 +89,117 @@ static FileLineColLoc findBestLocation(Location loc, bool emitted) {
   return {};
 }
 
+// Allow `json::Value`s to be used as map keys for the purpose of struct
+// definition uniquification. This abuses the `null` and `[null]` JSON values as
+// markers, and uses a very inefficient hashing of the value's JSON string.
+namespace llvm {
+template <>
+struct DenseMapInfo<JValue> {
+  static JValue getEmptyKey() { return nullptr; }
+  static JValue getTombstoneKey() { return JArray({nullptr}); }
+  static unsigned getHashValue(const JValue &x) {
+    SmallString<128> buffer;
+    llvm::raw_svector_ostream(buffer) << x;
+    return hash_value(buffer);
+  }
+  static bool isEqual(const JValue &a, const JValue &b) { return a == b; }
+};
+} // namespace llvm
+
 //===----------------------------------------------------------------------===//
 // HGLDD File Emission
 //===----------------------------------------------------------------------===//
 
 namespace {
 
+/// An emitted type.
+struct EmittedType {
+  StringRef name;
+  SmallVector<int64_t, 1> packedDims;
+  SmallVector<int64_t, 1> unpackedDims;
+
+  EmittedType() = default;
+  EmittedType(StringRef name) : name(name) {}
+  EmittedType(Type type) {
+    while (type) {
```

Comment:
while loops like this always make me concerned about infinite loops. in this case the loop is to descend potentially nested InOutTypes, right? i think it's safe, i just want to make sure i understand.

Reply:
Yes exactly. I don't think it's really needed though, because we only ever have one layer of InOutTypes that the HGLDD emission can make sense of. Which means that two calls to `getCanonicalType` and an optional inout unpacking in between should do the trick.

Let me change it to that simpler version. No need to have this loop here if it isn't exercised in practice.

Reply:
Makes sense, thanks.